File: SudokuSolver/Algorithms/CellFinderBySinglePossibleValue.cs
using SudokuSolver.Domain;

namespace SudokuSolver.Algorithms;

public class CellFinderBySinglePossibleValue : ICellFinder
{
    public CellFinderResult Find(HashSet<Cell> cells)
    {
        var cell = cells
            .Where(cell => !cell.IsOpened() && cell.PossibleValues.Count == 1)
            .FirstOrDefault();

        return cell is null
            ? CellFinderResult.CellNotFound()
            : CellFinderResult.CellFound(cell, cell.PossibleValues.First());
    }
}

File: src/main/java/ua/kh/poohdev/sudokusolver/algorithms/CellFinderBySinglePossibleValue.java
package ua.kh.poohdev.sudokusolver.algorithms;

import ua.kh.poohdev.sudokusolver.domain.Cell;

import java.util.stream.Stream;

public class CellFinderBySinglePossibleValue implements CellFinder {

    @Override
    public CellFinderResult find(Stream<Cell> cellsInUnit) {
        return cellsInUnit
                .filter(cell -> !cell.isOpened())
                .filter(cell -> cell.getPossibleValues().size() == 1)
                .map(cell -> CellFinderResult.found(cell, cell.getPossibleValues().iterator().next()))
                .findFirst()
                .orElseGet(CellFinderResult::notFound);
    }
}
